File: implementations/typescript/packages/api-server/src/redirect-state.ts
// SPDX-License-Identifier: Apache-2.0
import {record} from "@psp-cdl/core";
import {identifier} from "./service.js";

export const REDIRECT_PROFILE="PSP-LLM-REDIRECT-0.1";
/** Deliberately bounded profile grammar; never resolve this URI over the network. */
export function validRedirectTarget(value:unknown):value is string {
  return typeof value==="string"&&value.length<=256&&
    /^mcp:\/\/[a-z0-9]+(?:[.-][a-z0-9]+)*\/applications\/[A-Za-z0-9][A-Za-z0-9_-]*$/.test(value)&&!/[\r\n]/.test(value);
}
export function validRedirect(value:unknown):value is Record<string,unknown> {
  return record(value)&&Object.keys(value).sort().join(",")==="expiresAt,nodeId,nodeVersion,policyVersion,target"&&
    validRedirectTarget(value.target)&&[value.nodeId,value.nodeVersion,value.policyVersion].every(identifier)&&
    typeof value.expiresAt==="number"&&Number.isSafeInteger(value.expiresAt)&&value.expiresAt>0;
}

File: implementations/python/packages/api-server/src/psp_cdl_api_server/redirect_state.py
# SPDX-License-Identifier: Apache-2.0
"""Bounded host-only application targets; no network URI resolution."""
import re
from .service import identifier

REDIRECT_PROFILE = "PSP-LLM-REDIRECT-0.1"

def valid_redirect_target(value):
    return type(value) is str and len(value)<=256 and re.fullmatch(r"mcp://[a-z0-9]+(?:[.-][a-z0-9]+)*/applications/[A-Za-z0-9][A-Za-z0-9_-]*",value) is not None

def valid_redirect(value):
    return type(value) is dict and set(value)=={"expiresAt","nodeId","nodeVersion","policyVersion","target"} and valid_redirect_target(value["target"]) and all(identifier(value[k]) for k in ("nodeId","nodeVersion","policyVersion")) and type(value["expiresAt"]) in (int,float) and 0<value["expiresAt"]<=9007199254740991 and int(value["expiresAt"])==value["expiresAt"]
